modules/core/src/main/java/org/apache/ignite/internal/processors/service/IgniteServiceProcessor.java

```diff
@@ -1801,4 +1818,95 @@ private boolean enterBusy() {
     private void leaveBusy() {
         opsLock.readLock().unlock();
     }
+
+    /**
+     * Caches metrics to measure durations of service methods.
+     *
+     * @param srvc Service to measure durations of methods.
+     * @param srvcName Name of {@code srvc}.
+     */
+    private void cacheServiceMetrics(Service srvc, String srvcName) {
+        Map<GridServiceMethodReflectKey, HistogramMetricImpl> srvcHistograms = invocationHistograms.get(srvcName);
+
+        // The histogms might be already created by multipile-deploy.
+        if (srvcHistograms != null)
+            return;
+
+        srvcHistograms = new HashMap<>();
+
+        for (Class<?> itf : allInterfaces(srvc.getClass())) {
+            for (Method mtd : itf.getMethods()) {
+                if (metricIgnored(mtd.getDeclaringClass()))
+                    continue;
+
+                HistogramMetricImpl histogram = createHistogram(srvcName, mtd);
+
+                srvcHistograms.put(new GridServiceMethodReflectKey(mtd.getName(), mtd.getParameterTypes()), histogram);
+            }
+        }
+
+        invocationHistograms.put(srvcName, srvcHistograms);
```

Comment:
I suppose that invocationHistograms.put(srvcName, Collections.unmodifiableMap(srvcHistograms)); 
is even better.

Reply:
Fixed.